File: mandelbrot_examples/1_mandelbrot.cpp
#include <SFML/Graphics.hpp>
#include <complex>

const int WIDTH = 800;
const int HEIGHT = 600;
const int MAX_ITER = 1000;

sf::Color getColor(int iterations) {
    if (iterations == MAX_ITER) {
        return sf::Color::Black;
    }

    int r = (iterations * 9) % 256;
    int g = (iterations * 7) % 256;
    int b = (iterations * 5) % 256;
    return sf::Color(r, g, b);
}

int mandelbrot(const std::complex<double>& c) {
    std::complex<double> z = c;
    int n = 0;
    while (abs(z) <= 2.0 && n < MAX_ITER) {
        z = z * z + c;
        ++n;
    }
    return n;
}

void generateMandelbrot(sf::Image& image) {
    for (int x = 0; x < WIDTH; ++x) {
        for (int y = 0; y < HEIGHT; ++y) {
            double real = (x - WIDTH / 2.0) * 4.0 / WIDTH;
            double imag = (y - HEIGHT / 2.0) * 4.0 / HEIGHT;
            std::complex<double> c(real, imag);
            int iterations = mandelbrot(c);
            sf::Color color = getColor(iterations);
            image.setPixel(x, y, color);
        }
    }
}

int main() {
    sf::RenderWindow window(sf::VideoMode(WIDTH, HEIGHT), "Mandelbrot Set");
    sf::Image image;
    image.create(WIDTH, HEIGHT);
    generateMandelbrot(image);

    sf::Texture texture;
    texture.loadFromImage(image);
    sf::Sprite sprite(texture);

    while (window.isOpen()) {
        sf::Event event;
        while (window.pollEvent(event)) {
            if (event.type == sf::Event::Closed) {
                window.close();
            }
        }

        window.clear();
        window.draw(sprite);
        window.display();
    }

    return 0;
}

File: mandelbrot_examples/1_mandelbrot.py
# Cpp visualization of the Mandelbrot set
# Jesús Urtasun, Imperial College London - May 2024
# Visualizing the Mandelbrot set
print("Visualizing the Mandelbrot set")

import numpy as np
import matplotlib.pyplot as plt

def mandelbrot(c, max_iter):
    z = 0
    n = 0
    while abs(z) <= 2 and n < max_iter:
        z = z*z + c
        n += 1
    return n

# Image size (pixels)
width, height = 800, 800

# Plot window
re_min, re_max = -2.0, 1.0
im_min, im_max = -1.5, 1.5

# Maximum number of iterations
max_iter = 256

# Create a grid of complex numbers
re, im = np.linspace(re_min, re_max, width), np.linspace(im_min, im_max, height)
X, Y = np.meshgrid(re, im)
C = X + 1j * Y

# Compute the mandelbrot set
mandelbrot_set = np.frompyfunc(lambda c: mandelbrot(c, max_iter), 1, 1)(C).astype(np.float64)

# Plot the result
plt.figure(dpi=100)
plt.imshow(mandelbrot_set, extent=[re_min, re_max, im_min, im_max], cmap="hot")
plt.colorbar()
plt.title("Mandelbrot Set")
plt.xlabel("Re")
plt.ylabel("Im")
plt.show()